### ISSUE TITLE
no-unused-variable rule false positive on JSX elements in 3.4.0-dev.1 and TS 1.9.0-dev.20160216

### ISSUE THREAD

**Issue by myitcv** [palantir/tslint#972]
The following code (using `tslint 3.4.0-dev.1` and `tsc 1.9.0-dev.20160216`) gives a false positive in `no-unused-variable`:
 
```typescript
function BuildMyComponent(): JSX.Element {
	let StringComponentA = BuildComp();
	return <StringComponentA foo = { "test" } />;
}
```

```
[2, 6]: unused variable: 'StringComponentA'
```

If I find time later this week I can investigate further (unless anyone happens to beat me to it)

_Note that in order to repro this against `master` you will need to `npm install --save-dev typescript@next`_

On a related note, does it make sense to add a TypeScript nightly version to the Travis build matrix?

cc @adidahiya  @JKillian

**Comment by adidahiya** [comment]
We do build against `typescript@next` on the [next branch](https://github.com/palantir/tslint/blob/next/package.json#L42), for example [this PR](https://github.com/palantir/tslint/pull/961). Does `BuildComp()` return a stateless function component?

**Comment by myitcv** [comment]
> We do build against `typescript@next`

@adidahiya my apologies, should have checked first.

> Does `BuildComp()` return a stateless function component?

Hmm, yes all the instances where this is failing are stateless (at least I'm assuming stateless components are implicitly those with a state type of `{}`).

Have I missed something?

**Comment by adidahiya** [comment]
[SFCs](https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#stateless-function-components-in-react) are simple functions, not classes with constructor signatures. The compiler handles them a little differently. We currently don't have any tests for this new syntax.

**Comment by myitcv** [comment]
@adidahiya - I've reproduced this in TypeScript core and raised https://github.com/Microsoft/TypeScript/issues/7109

**Comment by JKillian** [comment]
@myitcv Do you think #1010 and #1034 are the same issues as this?

**Comment by myitcv** [comment]
@JKillian - potentially, but the bug that is listed in https://github.com/Microsoft/TypeScript/issues/7109 only appeared as of `1.9.0-dev.20160212`

https://github.com/palantir/tslint/issues/1034 could well be another issue because they user reports `v1.7.5`

I've asked the question about version information in https://github.com/palantir/tslint/issues/1010

Incidentally, I think it would be worth using the new GitHub [new issue template](https://github.com/blog/2111-issue-and-pull-request-templates) to encourage people to give us this sort of information up front...

**Comment by dallonf** [comment]
My ears are burning! Mentioned in detail in #1010, but worth repeating here: for me, this bug only shows up in Atom, and I believe the TSLint and TypeScript packages there track latest. So likely the same issue.

**Comment by myitcv** [comment]
@dallonf - I had forgotten that `atom-typescript` [behaves in this way](https://github.com/TypeStrong/atom-typescript/blob/master/docs/faq.md#which-version-of-typescript-does-atom-typescript-use). It's not helpful for sure... (@JKillian we could ask users whether they use Atom in the new issue template). Could well be the same issue therefore...

**Comment by JKillian** [comment]
Took me a minute to see what you were saying @myitcv: For you @dallonf, Atom TypeScript isn't working right since it's basically using a newer version of TypeScript than you're using, and the bug was introduced between those two versions.

This is unfortunate for sure, it's much nicer when bugs are in _older_ versions of TypeScript! Hopefully they can get this one fixed, bugs like this break reference finding and refactoring in IDEs, and IDE support is one major reason for using TS.

**Comment by myitcv** [comment]
@dallonf @JKillian @adidahiya - per [this comment](https://github.com/Microsoft/TypeScript/issues/7109#issuecomment-199685522), I believe this to have been fixed as of `1.9.0-dev.20160322`

**Comment by myitcv** [comment]
@adidahiya @JKillian - I closed this issue because I've successfully tested this as no longer broken... but please feel free to re-open if you think this issue is capturing something else?

**Comment by JKillian** [comment]
Thanks for the heads-up @myitcv! I tried it out as well and got the same results as you. Interestingly, #1034 also was fixed by this but not #1010 as far as I can tell.
